File: dotnet/Trinsic/AccountService.cs
using System;
using System.Threading.Tasks;
using Google.Protobuf;
using Microsoft.Extensions.Options;
using Okapi.Security;
using Okapi.Security.V1;
using Trinsic.Sdk.Options.V1;
using Trinsic.Services.Account.V1;

namespace Trinsic;

/// <summary>
/// Account Service
/// </summary>
public class AccountService : ServiceBase
{
    public AccountService(ServiceOptions options)
        : base(options) {
        Client = new(Channel);
    }

    public AccountService() {
        Client = new(Channel);
    }

    internal AccountService(ITokenProvider tokenProvider) : base(new(), tokenProvider) {
        Client = new(Channel);
    }

    internal AccountService(ITokenProvider tokenProvider, IOptions<ServiceOptions> options)
        : base(options.Value, tokenProvider) {
        Client = new(Channel);
    }
    
    /// <summary>
    /// Unprotects the account profile using a security code.
    /// The confirmation method field will specify how this code was
    /// communicated with the account owner.
    /// </summary>
    /// <param name="authToken"></param>
    /// <param name="securityCode"></param>
    public static string Unprotect(string authToken, string securityCode) {
        var profile = AccountProfile.Parser.ParseFrom(Base64Url.DecodeBytes(authToken));

        UnBlindOberonTokenRequest request = new() { Token = profile.AuthToken };
        request.Blinding.Add(ByteString.CopyFromUtf8(securityCode));
        var result = Oberon.UnblindToken(request);

        profile.AuthToken = result.Token;
        profile.Protection = new() {
            Enabled = false,
            Method = ConfirmationMethod.None
        };

        return Base64Url.Encode(profile.ToByteArray());
    }

    /// <summary>
    /// Protects the account profile with a security code.
    /// The code can be a PIN, password, keychain secret, etc.
    /// </summary>
    /// <param name="authToken"></param>
    /// <param name="securityCode"></param>
    public static string Protect(string authToken, string securityCode) {
        var profile = AccountProfile.Parser.ParseFrom(Base64Url.DecodeBytes(authToken));

        BlindOberonTokenRequest request = new() { Token = profile.AuthToken };
        request.Blinding.Add(ByteString.CopyFromUtf8(securityCode));
        var result = Oberon.BlindToken(request);

        profile.AuthToken = result.Token;
        profile.Protection = new() {
            Enabled = true,
            Method = ConfirmationMethod.Other
        };

        return Base64Url.Encode(profile.ToByteArray());
    }

    /// <summary>
    /// Gets the underlying grpc client
    /// </summary>
    private Account.AccountClient Client { get; }

    /// <summary>
    /// Logs in to the specified account; a new account will be created if it does not exist.
    /// </summary>
    /// <param name="request"></param>
    /// <returns></returns>
    public async Task<LoginResponse> LoginAsync(LoginRequest request) {
        if (string.IsNullOrWhiteSpace(request.EcosystemId))
            request.EcosystemId = DefaultEcosystem;

        var response = await Client.LoginAsync(request, await BuildMetadataAsync());

        if (response.ResponseCase == LoginResponse.ResponseOneofCase.Profile)
        {
            await TokenProvider.SaveAsync(Base64Url.Encode(response.Profile.ToByteArray()));
        }

        return response;
    }

    /// <summary>
    /// Logs in to the specified account; a new account will be created if it does not exist.
    /// </summary>
    /// <param name="request"></param>
    /// <returns></returns>
    public LoginResponse Login(LoginRequest request) {
        if (string.IsNullOrWhiteSpace(request.EcosystemId))
            request.EcosystemId = DefaultEcosystem;

        var response = Client.Login(request, BuildMetadata());

        if (response.ResponseCase == LoginResponse.ResponseOneofCase.Profile)
        {
            TokenProvider.Save(Base64Url.Encode(response.Profile.ToByteArray()));
        }

        return response;
    }

    /// <summary>
    /// Finalizes login from a previous `LoginRequest`.
    /// </summary>
    /// <param name="challenge">Challenge received from call to `Login`</param>
    /// <param name="authCode">Plaintext authentication code sent to account email</param>
    /// <returns>Auth token on success; null on failure</returns>
    public async Task<string?> LoginConfirmAsync(ByteString challenge, string authCode) {
        var hashed = Okapi.Hashing.Blake3.Hash(
            new() {
                Data = ByteString.CopyFromUtf8(authCode)
            }
        );

        var request = new LoginConfirmRequest() {
            Challenge = challenge,
            ConfirmationCodeHashed = hashed.Digest
        };

        var response = await Client.LoginConfirmAsync(request, await BuildMetadataAsync());

        if (response?.Profile == null)
            return null;

        var token = Base64Url.Encode(response.Profile.ToByteArray());

        // If profile is protected (it should be), unprotect it
        if (response.Profile.Protection?.Enabled ?? false)
            token = Unprotect(token, authCode);

        await TokenProvider.SaveAsync(token);
        return token;
    }

    /// <summary>
    /// Finalizes login from a previous `LoginRequest`.
    /// </summary>
    /// <param name="challenge">Challenge received from call to `Login`</param>
    /// <param name="authCode">Plaintext authentication code sent to account email</param>
    /// <returns>Auth token on success; null on failure</returns>
    public string? LoginConfirm(ByteString challenge, string authCode) {
        var hashed = Okapi.Hashing.Blake3.Hash(
            new() {
                Data = ByteString.CopyFromUtf8(authCode)
            }
        );

        var request = new LoginConfirmRequest() {
            Challenge = challenge,
            ConfirmationCodeHashed = hashed.Digest
        };

        var response = Client.LoginConfirm(request, BuildMetadata());

        if (response?.Profile == null)
            return null;

        var token = Base64Url.Encode(response.Profile.ToByteArray());

        // If profile is protected (it should be), unprotect it
        if (response.Profile.Protection?.Enabled ?? false)
            token = Unprotect(token, authCode);

        TokenProvider.Save(token);
        return token;
    }

    /// <summary>
    /// Creates an anonymous account in the current ecosystem
    /// </summary>
    /// <returns>Auth token for newly-created account; null on failure</returns>
    public string? LoginAnonymous(string ecosystemId = DefaultEcosystem) {
        var response = Login(new() { EcosystemId = ecosystemId });

        // `Profile` is returned on anonymous login only
        if (response.ResponseCase != LoginResponse.ResponseOneofCase.Profile)
            return null;

        // If profile is protected, this is clearly not an anonymous login
        if (response.Profile?.Protection?.Enabled ?? true)
            return null;

        var token = Base64Url.Encode(response.Profile.ToByteArray());
        TokenProvider.Save(token);

        return token;
    }

    /// <summary>
    /// Creates an anonymous account in the given ecosystem
    /// </summary>
    /// <returns>Auth token for newly-created account; null on failure</returns>
    public async Task<string?> LoginAnonymousAsync(string ecosystemId = DefaultEcosystem) {
        var response = await LoginAsync(new() { EcosystemId = ecosystemId });

        // `Profile` is returned on anonymous login only
        if (response.ResponseCase != LoginResponse.ResponseOneofCase.Profile)
            return null;

        // If profile is protected, this is clearly not an anonymous login
        if (response.Profile?.Protection?.Enabled ?? true)
            return null;

        var token = Base64Url.Encode(response.Profile.ToByteArray());
        await TokenProvider.SaveAsync(token);

        return token;
    }

    /// <summary>
    /// Return the details about the currently active account
    /// </summary>
    /// <returns></returns>
    public async Task<AccountInfoResponse> GetInfoAsync() {
        AccountInfoRequest request = new();
        var response = await Client.InfoAsync(new(), await BuildMetadataAsync(request));

        return response;
    }

    /// <summary>
    /// Return the details about the currently active account
    /// </summary>
    /// <returns></returns>
    public AccountInfoResponse GetInfo() {
        AccountInfoRequest request = new();
        var response = Client.Info(request, BuildMetadata(request));

        return response;
    }

// BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
// target: /home/runner/work/sdk/sdk/dotnet/Trinsic/AccountService.cs

	/// <summary>
    /// Get account information
    /// </summary>
    
    public AccountInfoResponse Info() {
        var request = new AccountInfoRequest();
        return Client.Info(request, BuildMetadata(request));
    }
	
	/// <summary>
    /// Get account information
    /// </summary>
    
    public async Task<AccountInfoResponse> InfoAsync() {
        var request = new AccountInfoRequest();
        return await Client.InfoAsync(request, await BuildMetadataAsync(request));
    }

	/// <summary>
    /// Authorize Ecosystem to receive webhook events
    /// </summary>
    
    public AuthorizeWebhookResponse AuthorizeWebhook(AuthorizeWebhookRequest request) {
        
        return Client.AuthorizeWebhook(request, BuildMetadata(request));
    }
	
	/// <summary>
    /// Authorize Ecosystem to receive webhook events
    /// </summary>
    
    public async Task<AuthorizeWebhookResponse> AuthorizeWebhookAsync(AuthorizeWebhookRequest request) {
        
        return await Client.AuthorizeWebhookAsync(request, await BuildMetadataAsync(request));
    }

// END Code generated by protoc-gen-trinsic. DO NOT EDIT.
}

File: python/trinsic/account_service.py
import base64
import string
from typing import Union, SupportsBytes

from trinsicokapi import oberon, hashing
from trinsicokapi.proto.okapi.hashing.v1 import Blake3HashRequest
from trinsicokapi.proto.okapi.security.v1 import (
    UnBlindOberonTokenRequest,
    BlindOberonTokenRequest,
)

from trinsic.proto.sdk.options.v1 import ServiceOptions
from trinsic.proto.services.account.v1 import *
from trinsic.security_providers import ITokenProvider
from trinsic.service_base import ServiceBase


class AccountService(ServiceBase):
    """Wrapper for the [Account Service](/reference/services/account-service/)"""

    def __init__(
        self,
        server_config: ServiceOptions = None,
        token_provider: ITokenProvider = None,
    ):
        """
        Initialize a connection to the server.
        Args:
            server_config: The URL of the server, or a channel which encapsulates the connection already.
        """
        super().__init__(server_config, token_provider)
        self.client = AccountStub(super().channel)

    async def sign_in(self, *, request: SignInRequest = None) -> str:
        """
        Deprecated - use `AccountService.login` and `AccountService.login_confirm` instead
        """
        request = request or SignInRequest()
        request.ecosystem_id = request.ecosystem_id or "default"
        response = await self.client.sign_in(
            sign_in_request=request, metadata=self.build_metadata()
        )
        auth_token = base64.urlsafe_b64encode(bytes(response.profile)).decode("utf-8")
        self.token_provider.save(auth_token)
        return auth_token

    @staticmethod
    def unprotect(
        *,
        profile: Union[AccountProfile, str],
        security_code: Union[SupportsBytes, bytes, str],
    ) -> str:
        """
        Unprotect the account profile using a security code. The confirmation method field will specify how this code was communicated with the account owner.
        Args:
            profile:
            security_code:
        Returns:
            The in-place unprotected `AccountProfile`
        """
        if isinstance(profile, str):
            profile = AccountProfile().parse(base64.urlsafe_b64decode(profile))
        request = UnBlindOberonTokenRequest(token=profile.auth_token)
        if isinstance(security_code, str):
            security_code = bytes(security_code, "utf8")
        request.blinding.append(bytes(security_code))
        result = oberon.unblind_token(request)
        profile.auth_token = result.token
        profile.protection.enabled = False
        profile.protection.method = ConfirmationMethod.None_
        return base64.urlsafe_b64encode(bytes(profile)).decode("utf-8")

    @staticmethod
    def protect(
        *,
        profile: Union[AccountProfile, str],
        security_code: Union[SupportsBytes, bytes, str],
    ) -> str:
        """
        Protect the account profile with a security code. The code can be a PIN, password, keychain secret, etc.
        Args:
            profile:
            security_code:
        Returns:
            A protected `AccountProfile`
        """
        if isinstance(profile, str):
            profile = AccountProfile().parse(base64.urlsafe_b64decode(profile))
        request = BlindOberonTokenRequest(token=profile.auth_token)
        request.blinding.append(bytes(security_code))
        result = oberon.blind_token(request)
        profile.auth_token = result.token
        profile.protection.enabled = True
        profile.protection.method = ConfirmationMethod.Other
        return base64.urlsafe_b64encode(bytes(profile)).decode("utf-8")

    async def login(self, *, request: LoginRequest) -> LoginResponse:
        """
        Begins the login process for the specified user; returning a challenge which must be passed
        to `login_confirm()` along with an auth code sent to the user's email
        Args:
            request: Login request
        Returns:
            LoginResponse with challenge
        """
        request = request or LoginRequest()
        response = await self.client.login(request, metadata=self.build_metadata())
        if response.profile:
            self.token_provider.save(
                base64.urlsafe_b64encode(bytes(response.profile)).decode("utf8")
            )
        return response

    async def login_confirm(self, *, challenge: bytes, auth_code: str) -> str:
        """
        Finalizes the login process using the challenge received from `login()`,
        and auth code sent to user's email
        Args:
            challenge: The challenge contained in the response to `login()`
            auth_code: The auth code sent to the user's email
        Returns:
            Authentication token for account
        """

        code_hash = hashing.blake3_hash(
            request=Blake3HashRequest(data=auth_code.encode("utf-8"))
        ).digest

        request = LoginConfirmRequest(
            challenge=challenge, confirmation_code_hashed=code_hash
        )
        response = await self.client.login_confirm(
            request, metadata=self.build_metadata()
        )

        auth_token = base64.urlsafe_b64encode(bytes(response.profile)).decode("utf-8")

        if response.profile.protection.enabled:
            auth_token = self.unprotect(profile=auth_token, security_code=auth_code)

        self.token_provider.save(auth_token)
        return auth_token

    async def login_anonymous(self, *, ecosystem_id: string = None) -> string:
        """
        Create an anonymous account in the given ecosystem
        Args:
            ecosystem_id: ID of ecosystem to sign into; defaults to "default" if unspecified
        Returns:
            Authentication token for newly-created account
        """

        ecosystem_id = ecosystem_id or ""

        response = await self.login(request=LoginRequest(ecosystem_id=ecosystem_id))
        auth_token = base64.urlsafe_b64encode(bytes(response.profile)).decode("utf-8")

        self.token_provider.save(auth_token)

        return auth_token

    async def get_info(self) -> AccountInfoResponse:
        """
        Return the details about the currently active account.
        Returns:
            The `InfoResponse`
        """
        return await self.info()

    # BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
    # target: /home/runner/work/sdk/sdk/python/trinsic/account_service.py

    async def info(self) -> AccountInfoResponse:
        """Get account information"""
        request = AccountInfoRequest()
        return await self.client.info(request, metadata=self.build_metadata(request))


# END Code generated by protoc-gen-trinsic. DO NOT EDIT.
